src/scanner.c

```diff
@@ -149,63 +151,98 @@ static Scanner scanners[] = {
 	}
 };
 
-/*
- * Perform either a heap or index scan depending on the information in the
- * ScannerCtx. ScannerCtx must be setup by caller with the proper information
- * for the scan, including filters and callbacks for found tuples.
- *
- * Return the number of tuples that where found.
- */
-int
-ts_scanner_scan(ScannerCtx *ctx)
+static inline Scanner *
+scanner_ctx_get_scanner(ScannerCtx *ctx)
+{
+	if (OidIsValid(ctx->index))
+		return &scanners[ScannerTypeIndex];
+	else
+		return &scanners[ScannerTypeHeap];
+}
+
+void
+ts_scan_iterator_init(ScanIterator *iterator)
 {
 	TupleDesc tuple_desc;
-	bool is_valid;
 	Scanner *scanner;
 
-	InternalScannerCtx ictx = {
-		.sctx = ctx,
-	};
+	Assert(NULL == iterator->tinfo);
 
-	if (OidIsValid(ctx->index))
-		scanner = &scanners[ScannerTypeIndex];
-	else
-		scanner = &scanners[ScannerTypeHeap];
+	if (iterator->ictx == NULL)
+		iterator->ictx = palloc0(sizeof(*iterator->ictx));
+
+	iterator->ictx->sctx = &iterator->ctx;
+	iterator->ictx->closed = false;
 
-	scanner->openheap(&ictx);
-	scanner->beginscan(&ictx);
+	iterator->tinfo = &iterator->ictx->tinfo;
 
-	tuple_desc = RelationGetDescr(ictx.tablerel);
+	scanner = scanner_ctx_get_scanner(&iterator->ctx);
 
-	ictx.tinfo.scanrel = ictx.tablerel;
-	ictx.tinfo.desc = tuple_desc;
-	ictx.tinfo.mctx = ctx->result_mctx == NULL ? CurrentMemoryContext : ctx->result_mctx;
+	scanner->openheap(iterator->ictx);
+	scanner->beginscan(iterator->ictx);
+
+	tuple_desc = RelationGetDescr(iterator->ictx->tablerel);
+
+	iterator->ictx->tinfo.scanrel = iterator->ictx->tablerel;
+	iterator->ictx->tinfo.desc = tuple_desc;
+	iterator->ictx->tinfo.mctx =
+		iterator->ctx.result_mctx == NULL ? CurrentMemoryContext : iterator->ctx.result_mctx;
 
 	/* Call pre-scan handler, if any. */
-	if (ctx->prescan != NULL)
-		ctx->prescan(ctx->data);
+	if (iterator->ctx.prescan != NULL)
+		iterator->ctx.prescan(iterator->ctx.data);
+}
+
+static inline bool
+ts_scanner_limit_reached(ScannerCtx *ctx, ScanIterator *iterator)
+{
+	return ctx->limit > 0 && iterator->ictx->tinfo.count >= ctx->limit;
+}
+
+void
+ts_scan_iterator_close(ScanIterator *iterator)
+{
+	Scanner *scanner = scanner_ctx_get_scanner(iterator->ictx->sctx);
+
+	if (iterator->ictx->closed)
+		return;
+
+	/* Call post-scan handler, if any. */
+	if (iterator->ictx->sctx->postscan != NULL)
+		iterator->ictx->sctx->postscan(iterator->ictx->tinfo.count, iterator->ictx->sctx->data);
 
-	is_valid = scanner->getnext(&ictx);
+	scanner->endscan(iterator->ictx);
+	scanner->closeheap(iterator->ictx);
+	iterator->ictx->closed = true;
+}
+
+bool
+ts_scan_iterator_next(ScanIterator *iterator)
```

Comment:
I renamed it as  `_advance`. I think it's slightly easier to use if you don't need to declare the TupleInfo in the calling code. Will change if you feel strongly. (also other scanner getnext functions return a bool not a pointer)

Reply:
There's not a huge difference between returning `bool` and a pointer given that pointers can be evaluated as booleans. So, even just changing this function to returning a pointer (I guess it would return `TupleInfo *` and not `Tuple *`) would be possible without changing the foreach interface. Then, at least, one has the option to forgo the `foreach` macro and do a plain `for` loop that gives access to the `TupleInfo *`.

I just find it a bit unclear from the `foreach` interface how you actually access the iterated elements. It is somewhat awkward that you have to reach into the internals of the iterator to grab the `TupleInfo` when it does a good job otherwise hiding the internal details. At the very least, we should probably have some accessor functions/macros `Tuple *tup = ts_scan_iterator_tuple(&iterator)` and `TupleDesc desc = ts_scan_iterator_tupledesc(&iterator)` that make the interface explicit and avoid having to reach into multiple levels of nested structs (e.g., `iterator.tinfo->tuple`). So, I suggest that we:

* Change this function to return `TupleInfo *` while keeping the `foreach` macro the same (in that case `_next` is a suitable function name suffix).
* Add accessors for `Tuple` and `TupleDesc`